[FIX] website, website_sale, sale: add to cart + checkout mobile fixes

dabo-odoo opened the pull request:
Before the change, in the mobile version of the eCommerce cart and
add to cart page, if a product had a name that was too long or the
font size was too big it would push the elements to its right
offscreen.
After the change, the overflow of a product name that is too long or
too big gets replaced with ellipsis, keeping the elements to its right
from beign pushed offscreen.

The add to cart mobile page has also been refactored to be more
similar to the checkout page.

Furthermore, a bug with the alert message that caused it to be
squished in the "add to cart" page and a bug in the same page with
the background scrollbar still beign visible when the dialog is open
have been fixed.

task-4260720
opw-4253880

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/188842.png)](https://mergebot.odoo.com/odoo/odoo/pull/188842)

ltinel commented:
Hi @anso-odoo, could I have your opinion/review regarding these changes? :slightly_smiling_face: Thanks in advance!

anso-odoo commented:
Hello @ltinel @dabo-odoo 

IMO we can't do miracles with this modal, because the use of tables makes every layout change very difficult/not optimised

I didn't check the code but visually, what you already did is moslty fine. I have some suggestions though: 

- IMO the "Quantity" heading can just go away in mobile. It brings nothing and is placed above both the input and the price, which does not make a lot of sense
- `.css_quantity` is not aligned with the end of the table. A combination of `justify-content-end justify-content-md-center` should do the trick
- Not sure if it's possible, but I think it would make sense to put the price on top, as it's a bit floating where it is right now, and it is more important than the quantity field IMO

What gives us: 
<img width="443" alt="Screenshot 2024-12-05 at 15 12 59" src="https://github.com/user-attachments/assets/508000ed-5209-41f8-ade2-983a3ec96182">

WDYT ? (@lebl-odoo too)

Note:
Ideally, I think this would be a more optimal layout, but it would involve using flex instead of tables 🤓:
![Frame 917](https://github.com/user-attachments/assets/f9401bf6-5210-43ca-8515-9fc3c07d6d7f)

lebl-odoo commented:
LGTM @anso-odoo , yes, flex would be ideal. Not sure if it's realistic in this PR. Let me know @dabo-odoo :)

dabo-odoo commented:
@Itinel @lebl-odoo @anso-odoo
When I first started this PR my first idea was revamping the layout using either more rows or changing to flex, but doing so broke a lot of tours and so I chose the safest option of simply adjusting the existing layout.
If you want me to implement the layout above I will need to change a lot of tours and possibly tests, let me know if it's safe and ok for you.

lebl-odoo commented:
I can live with it, but I let the design team the last word, @dabo-odoo @dabo-odoo

anso-odoo commented:
@lebl-odoo I'm sorry, my note was more a rant against the fact that we are using tables here and I think that we should consider to change it one day or another. That said, I realise it's clearly OOS for this PR and the result is mostly fine, so I can live without too 👍

dabo-odoo commented:
@lebl-odoo @anso-odoo @ltinel
I kept the table layout, refactored using bootstrap and implemented the requested changes, let me know if it's ok
https://runbot.odoo.com/runbot/batch/1725537

dabo-odoo commented on addons/sale/static/src/js/product_list/product_list.scss:
```diff
@@ -0,0 +1,25 @@
+.fixed_border_for_block_display {
+    border-bottom: solid;
+    border-bottom-width: 1px;
+    border-color: inherit;
```
I had to override the tr styling because it would appear like this otherwise
![2024-12-10_09-53](https://github.com/user-attachments/assets/1b06c4b8-f0ff-4ba9-8a87-18f10f619ce7)

lebl-odoo commented:
All good to me @dabo-odoo

ltinel commented on addons/sale/static/src/js/product_list/product_list.scss:
```diff
@@ -0,0 +1,25 @@
+}
+
+@include media-breakpoint-down(md) {
+    .fixed_border_for_block_display {
```
Could you use a clearer class name or select the element without adding a class?

ltinel reviewed (changes requested):
A few more comments :) I'll have a look at your runbot tomorrow, to see the changes in action :)

dabo-odoo commented on addons/sale/static/src/js/product/product.xml:
```diff
@@ -27,49 +24,52 @@
-                    t-att-class="{'disabled': !this.env.isPossibleCombination(this.props)}"
-                    t-on-click="() => this.env.addProduct(this.props.product_tmpl_id)"
+                <td
+                    class="o_sale_product_configurator_qty py-3 px-0 text-end text-md-center"
```
Yes, I already tried and a lot of tests failed, I would prefer creating another PR to address this since this one is already OOS in my opinion.

ltinel commented on addons/sale/static/src/js/product/product.scss:
```diff
@@ -17,18 +17,23 @@
-
-    @include media-breakpoint-up(md) {
+@include media-breakpoint-up(md) {
+    .o_sale_product_configurator_qty {
+        width: 160px;
+    }
+    
+    .o_sale_product_configurator_price {
```
Nit:
```suggestion
    .o_sale_product_configurator_qty,
    .o_sale_product_configurator_price {
```

ltinel commented on addons/website_sale/views/templates.xml:
```diff
@@ -1911,25 +1911,27 @@
+                    <div class="flex-grow-1 overflow-hidden">
                         <t t-call="website_sale.cart_line_product_link">
-                            <h6 t-field="line.name_short" class="d-inline align-top h6 fw-bold"/>
+                            <h6 t-field="line.name_short" class="d-block align-top h6 fw-bold overflow-hidden text-truncate"/>
```
```suggestion
                            <h6 t-field="line.name_short" class="d-block align-top h6 fw-bold text-truncate"/>
```

ltinel commented on addons/sale/static/src/js/product/product.xml:
```diff
@@ -27,49 +24,58 @@
+                    <a
+                        class="d-block mt-2 text-end"
+                        role="button"
+                        t-if="!this.props.optional &amp;&amp; !isMainProduct"
```
No need to check `this.props.optional` again here :)
```suggestion
                        t-if="!isMainProduct"
```

ltinel commented on addons/sale/static/src/js/product/product.xml:
```diff
@@ -27,49 +24,58 @@
+                </td>
+                <td
+                    class="o_sale_product_configurator_price py-3 px-0 text-end text-md-center"
+                    name="price"
```
I'm not sure whether it's a good practice to use the same `name` attribute twice. I guess there's no harm if it doesn't cause any issues?

dabo-odoo replied:
They are mutually exclusive (if - else) so it will not be a problem, but this whole section should be refactored in a separate PR as you told me before IMO.

ltinel commented on addons/website/static/src/scss/website.ui.scss:
```diff
@@ -146,4 +146,10 @@ body {
     }
+    
+    @include media-breakpoint-down(md) {
+        &:has(.modal-open) {
```
This doesn't seem to be applied. Could you try without the `&`? Or maybe something like `body.modal-open` will do the trick?

dabo-odoo replied:
I tested it now and I see it working, I do not know if you meant somewhere else
![modal-open](https://github.com/user-attachments/assets/4f3421c4-3516-4405-8366-47b7a865365a)

ltinel replied:
In Odoo's mobile preview, it seems to work, but if you test this by simply reducing the screen width (or using the mobile previous from the Chrome DevTools), it doesn't seem to work :/

qsm-odoo replied:
That's probably not what we want: overflow: hidden is already added by bootstrap, but not through the modal-open class anymore. If we are to restore bootstrap to do what it did in a distant past, there would need to be a very convincing comment on top of that CSS you are adding.
![image](https://github.com/user-attachments/assets/7b52335a-af55-49b3-b148-1b8c36d32981)

We also have JS logic about opening/closing dialog blocking the scroll *or not* depending on the presence of a backdrop or not. Touching the modal-open class behavior seems risky.

ltinel reviewed (approved):
Thanks, looks great! :star_struck: I wonder whether we should also reduce the padding of the quantity `input` in mobile (same as for the `+` and `-` buttons). WDYT?

Sorry for the delay BTW, I took a lot of time off over the last week :sweat_smile:

dabo-odoo commented:
@ltinel 
I tried reducing the padding and it ended up looking better IMO, let me know if anything else needs to be done
https://runbot.odoo.com/runbot/batch/1734071

ltinel commented on addons/website/static/src/scss/website.ui.scss:
```diff
@@ -147,3 +147,7 @@ body {
     }
 }
+
+body.modal-open:has(.js_sale.o_wsale_product_page) {
```
This style only targets the product page, so it should probably be moved to `website_sale` :)
